core/trino-main/src/main/java/io/trino/execution/resourcegroups/InternalResourceGroup.java

```diff
@@ -785,37 +784,31 @@ private ResourceUsage updateResourceUsageAndGetDelta()
         synchronized (root) {
             ResourceUsage groupUsageDelta = new ResourceUsage(0, 0);
 
-            if (subGroups.isEmpty()) {
-                // Leaf resource group
-                for (Map.Entry<ManagedQueryExecution, ResourceUsage> entry : runningQueries.entrySet()) {
-                    ManagedQueryExecution query = entry.getKey();
-                    ResourceUsage oldResourceUsage = entry.getValue();
-
-                    ResourceUsage newResourceUsage = new ResourceUsage(
-                            query.getTotalCpuTime().toMillis(),
-                            query.getTotalMemoryReservation().toBytes());
-
-                    // Compute delta and update usage
-                    ResourceUsage queryUsageDelta = newResourceUsage.subtract(oldResourceUsage);
-                    entry.setValue(newResourceUsage);
-                    groupUsageDelta = groupUsageDelta.add(queryUsageDelta);
-                }
+            for (Map.Entry<ManagedQueryExecution, ResourceUsage> entry : runningQueries.entrySet()) {
+                ManagedQueryExecution query = entry.getKey();
+                ResourceUsage oldResourceUsage = entry.getValue();
 
-                cachedResourceUsage = cachedResourceUsage.add(groupUsageDelta);
+                ResourceUsage newResourceUsage = new ResourceUsage(
+                        query.getTotalCpuTime().toMillis(),
+                        query.getTotalMemoryReservation().toBytes());
+
+                // Compute delta and update usage
+                ResourceUsage queryUsageDelta = newResourceUsage.subtract(oldResourceUsage);
+                entry.setValue(newResourceUsage);
+                groupUsageDelta = groupUsageDelta.add(queryUsageDelta);
             }
-            else {
-                // Intermediate resource group
-                for (InternalResourceGroup subGroup : dirtySubGroups) {
-                    ResourceUsage subGroupUsageDelta = subGroup.updateResourceUsageAndGetDelta();
-                    groupUsageDelta = groupUsageDelta.add(subGroupUsageDelta);
-                    cachedResourceUsage = cachedResourceUsage.add(subGroupUsageDelta);
-
-                    if (!subGroupUsageDelta.equals(new ResourceUsage(0, 0))) {
-                        subGroup.updateEligibility();
-                    }
+
+            for (InternalResourceGroup subGroup : dirtySubGroups) {
+                ResourceUsage subGroupUsageDelta = subGroup.updateResourceUsageAndGetDelta();
+                groupUsageDelta = groupUsageDelta.add(subGroupUsageDelta);
+
+                if (!subGroupUsageDelta.equals(new ResourceUsage(0, 0))) {
+                    subGroup.updateEligibility();
                 }
             }
 
+            cachedResourceUsage = cachedResourceUsage.add(groupUsageDelta);
```

Comment:
For the intermediate group case, this was updated before the call to `subGroup.updateEligibility();` which calls `parent.get().updateEligibility` and uses `cachedResourceUsage`.
It would be better to keep the original order here unless you are sure it does not matter

Reply:
My understanding is that updating this field before every call to `subGroup.updateEligibility()` was unnecessary (I think I’d need to draw it to explain why clearly 😉). Just in case I'm missing something, I reverted it to the previous version.